common/src/main/java/org/jboss/pnc/common/gerrit/ScmUrlGeneratorProvider.java

```diff
@@ -0,0 +1,131 @@
+/**
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2014-2022 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jboss.pnc.common.gerrit;
+
+import org.jboss.pnc.common.net.GitSCPUrl;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.validation.constraints.NotNull;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Map;
+
+/**
+ * Class to deal with Gerrit/Gitlab information
+ */
+@ApplicationScoped
+public class ScmUrlGeneratorProvider {
+
+    public enum SCMProvider {
+        GERRIT, GITLAB
+    }
+
+    private static final Map<SCMProvider, ScmUrlGenerator> scmUrlProviders = Map
+            .of(SCMProvider.GERRIT, new GerritScmUrlGenerator(), SCMProvider.GITLAB, new GitlabScmUrlGenerator());
+
+    /**
+     * Returns an SCM url generator specific to a type of SCM
+     *
+     * @param provider type of SCM provider (Gerrit, Gitlab)
+     * @return generator of urls specific to the type of provider
+     * @throws ScmException throws if provider is unknown
+     */
+    public ScmUrlGenerator getScmUrlGenerator(@NotNull SCMProvider provider) throws ScmException {
+        var generator = scmUrlProviders.get(provider);
+        if (generator == null) {
+            throw new ScmException("Unknown SCM provider: " + provider);
+        }
+        return generator;
+    }
+
+    /**
+     * Simple heuristic logic for trying to determine what is the SCM provider out of url used to clone the build and
+     * internal representation of scm repo from the SCMRepository entity.
+     *
+     * @param scmUrl url used to clone repo in a Build (URI only)
+     * @param internalScmUrl internalUrl from SCMRepository of the Build (can be URI or SCP-like)
+     * @return SCMProvider that is likely to have been used
+     * @throws ScmException if the arguments are not appropriate URLs
+     */
+    public SCMProvider determineScmProvider(String scmUrl, String internalScmUrl) throws ScmException {
+        URI scmUri;
+        try {
+            scmUri = new URI(scmUrl);
```

Comment:
Couldn't this be a problem? Gitlab uses the same format.
```
jshell> new URI("git@github.com:project-ncl/pnc.git");
|  Exception java.net.URISyntaxException: Illegal character in scheme name at index 3: git@github.com:project-ncl/pnc.git
|        at URI$Parser.fail (URI.java:2976)
|        at URI$Parser.checkChars (URI.java:3147)
|        at URI$Parser.parse (URI.java:3173)
|        at URI.<init> (URI.java:623)
|        at (#1:1)
```

Reply:
the `scmUrl` which is from `BuildRecord.scmRevisionUrl` is should in form of URI (from what I've seen in stage), so there shouldn't be a case where it is scp-like